### DISCUSSION TITLE
Issue using JDBI3 and Iterator in Dropwizard version 2.1.5

### DISCUSSION THREAD

**Discussion by irineuruiz** [dropwizard/dropwizard#7175] [General]
Hi,

We tried to update Dropwizard from version 2.1.4 to 2.1.5 and ran into an issue that didn't occur before.
We have an interface using JDBI3 with a simple query that returns an iterator and may or may not have results.

An example:

```
@RegisterRowMapper(DataMapper.class)
public interface DataDAO {

    @SqlQuery("a simple select with parameters")
    Iterator<Data> getDataInTime(@Bind("beginData") Timestamp begin,
            @Bind("endData") Timestamp end);


}

```

And a resource that tries to get the data from that query and shows on the screen if it came empty or not

```
    @GET
    @Path("test")
    public String test() {
        DateTime beginData = new DateTime();
        DateTime endData = new DateTime();
        Iterator<Data> result = daoTeste.getDataInTime(
                new Timestamp(beginData.getMillis()),
                new Timestamp(endData.getMillis()));
        if (!result.hasNext()) {
            return "Empty";
        }
        return "With data";
    }

```

When we run this test, it shouldn't find anything in our example database and return "Empty" as a response and it works as expected with Dropwizard 2.1.4.
However, when migrating to version 2.1.5, it starts to give an error due to the ResultSet being closed.

```
! org.postgresql.util.PSQLException: Este ResultSet está fechado.
! at org.postgresql.jdbc.PgResultSet.checkClosed(PgResultSet.java:3095)
! at org.postgresql.jdbc.PgResultSet.next(PgResultSet.java:2107)
! at org.jdbi.v3.core.result.ResultSetResultIterator.safeNext(ResultSetResultIterator.java:104)
! ... 99 common frames omitted
! Causing: org.jdbi.v3.core.result.ResultSetException: Unable to advance result set [statement:"***************************************************************************************************"]
! at org.jdbi.v3.core.result.ResultSetResultIterator.safeNext(ResultSetResultIterator.java:106)
! at org.jdbi.v3.core.result.ResultSetResultIterator.hasNext(ResultSetResultIterator.java:58)
! at br.com.rassystem.resources.v1.frota.FrotaResource.teste(FrotaResource.java:106)
! at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
! at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
! at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
! at java.base/java.lang.reflect.Method.invoke(Method.java:568)
! at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory.lambda$static$0(ResourceMethodInvocationHandlerFactory.java:52)
! at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:134)
! at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:177)
! at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$TypeOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:219)
! at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:81)
! at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:478)
! at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:400)
! at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:81)
! at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:256)

```
We saw that the cause is the jdbi3 version change from 3.34.0 to 3.37.1, but I am in doubt if what we are doing is really wrong or if this is a bug.
Can't we use iterator in jdbi3 with dropwizard anymore?

**Comment by joschi** [comment]
@irineuruiz Thanks for reporting!

Could you please try updating _only_ the jdbi dependencies in your project and not Dropwizard, and then try again?

**Comment by irineuruiz** [comment]
Olá @joschi,

I already did that and from version 3.35.0 it starts to give this error.

But I wanted to talk to you first because in the jdbi 3 documentation there is a text about asynchronous execution

```
https://jdbi.org/#_usage_in_asynchronous_applications
"Note, since the handle closes as soon as the callback returns, you cannot return an iterator, since iteration will call upon the (now closed) handle"_
```

So I don't know if this is a bug in jdbi3 or if it has to do with how this is implemented internally in dropwizard

**Comment by joschi** [comment]
There were quite some changes between jdbi 3.34.0 and 3.35.0:
https://github.com/jdbi/jdbi/compare/v3.34.0...v3.35.0

Could you please create small reproducer for this issue and file a bug report with the jdbi project?
https://github.com/jdbi/jdbi/issues

> But I wanted to talk to you first because in the jdbi 3 documentation there is a text about asynchronous execution

As far as I see you're not using any async code in the snippets you've been posting and there's also nothing pointing to this in your stack trace.

**Comment by irineuruiz** [comment]
[jdbiIterator.zip](https://github.com/dropwizard/dropwizard/files/11019008/jdbiIterator.zip)

In this zip I have a simple case with the problem using JDBI3 and dropwizard.

I'll try to mount the same scenario without the dropwizard and check if the problem also occurs to report to the JDBI3 staff.

I will keep you informed

CREATE TABLE public.tb_person
(
id integer NOT NULL,
firstname text,
dt_insert timestamp with time zone,
CONSTRAINT pk00_tb_person PRIMARY KEY (id)
)
WITH (
OIDS=FALSE
);

**Comment by irineuruiz** [comment]
I found an open issue on the jdbi3 side since January of this year.

I posted this example there too

https://github.com/jdbi/jdbi/issues/2255

**Comment by hgschmie** [comment]
Hi @joschi , @irineuruiz. I am one of the contributors to Jdbi. Correct, this is a behavior change. Let me look into this a bit more and I will give you an update.

**Comment by hgschmie** [comment]
I added a longer chapter (http://jdbi.org/#using-steams-and-iterators) to the Jdbi documentation, which will be in the developer documentation already and be part of the 3.38.0 release documentation. It is understood that the behavior change is annoying, but the code never worked correctly before (it was always prone to leaking resource in some corner cases) and the changes in 3.35.0 actually fixed those.
